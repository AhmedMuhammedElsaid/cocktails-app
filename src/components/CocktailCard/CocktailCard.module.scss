@import "../../scss//variables.scss";

.Card {
  text-decoration: none;
  min-height: 360px;
  max-width: 345px;
  margin: 5px auto;
  min-width: 355px;
  color: $darkGrey;
  box-shadow: 5px 5px 5px 0px $boxShadow;
  transition: $mainTransition;
  button {
    min-height: 100%;
  }
  &:hover {
    box-shadow: 6px 7px 5px 0px $darkBoxShadow;
    color:$mainOrange;
    text-decoration:underline;
    cursor: pointer;
    opacity: .8;
  }
  &__Img {
    padding: 1em;
    object-fit: contain !important;
  }
}
