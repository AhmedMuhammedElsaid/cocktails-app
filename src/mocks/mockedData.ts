export const cocktailDetailsMockedData = {
  idDrink: "14608",
  strDrink: "Cream Soda",
  strCategory: "Cocktail",
  strAlcoholic: "Alcoholic",
  strGlass: "Highball glass",
  strInstructions:
    "Pour 1oz of Spiced Rum into a highball glass with ice. Fill with Ginger Ale.",
  strInstructionsDE:
    "Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale .",
  strInstructionsIT: `Versare 30gr di rum speziato in un bicchiere highball con ghiaccio.Riempi con Ginger Ale.`,
  strDrinkThumb: `https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg`,
  strIngredient1: `Spiced rum`,
  strIngredient2: `Ginger ale`,
  strCreativeCommonsConfirmed: `No`,
  dateModified: `2016-11-15 11:29:27`,
};
export const cocktailListMockedData = [
  cocktailDetailsMockedData,
  { ...cocktailDetailsMockedData, idDrink: "14709" },
];
export const cocktailCardMockedData = {
  strDrink: "Cream Soda",
  strCategory: "Cocktail",
  strDrinkThumb: `https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg`,
  strInstructions:
    "Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale .",
};
