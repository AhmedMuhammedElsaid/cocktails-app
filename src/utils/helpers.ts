/*
 @params object
 return object with no null ,undefined , empty string "cleaned object"
*/
export function cleanObject(object: { [x: string]: any }) {
  if (!object) return {};
  if (typeof object === "object") {
    for (let propName in object) {
      if (
        object[propName] === null ||
        object[propName] === undefined ||
        object[propName]?.length === 0
      ) {
        delete object[propName];
      }
    }
  }
  return object;
}
