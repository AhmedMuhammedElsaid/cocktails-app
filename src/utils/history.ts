import { createBrowserHistory } from "history";

const history = createBrowserHistory();
// method to redirect the user to error page in case any error happens while fetching the data from the server
export const redirectToErrorPage = () => {
  history.push(`/pagenotfound`);
  history.go(0);
};
export default history;
