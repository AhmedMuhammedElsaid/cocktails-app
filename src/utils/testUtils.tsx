import { render, RenderResult } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const customRender = (component: JSX.Element): RenderResult => {
  return {
    ...render(<Router>{component}</Router>),
  };
};
