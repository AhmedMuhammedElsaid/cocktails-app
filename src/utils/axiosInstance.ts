import axios from "axios";
import { redirectToErrorPage } from './history';
import { toast } from "react-toastify";

const axiosInstance = axios.create({
  // setting the default uri for this instance
  baseURL: process.env.REACT_APP_BASE_URI,
});

// setting language  configuration on every single request we send to the BE
axiosInstance.interceptors.request.use(
  (request: any) => {
    return {
      ...request,
      headers: {
        ...request.headers,
        // setting the header lang german "de"
        "Accept-Language": "de",
      },
    };
  },
  (error) => {
    return Promise.reject(error);
  }
);

axiosInstance.interceptors.response.use(
  (response) => {
    return response.data;
  },
  (error) => {
    switch (error?.response?.status) {
      case 400:
        return {
          data: null,
          hasError: true,
          error: [error.response.data],
        };
      case 401:
        toast.error("Unauthorized Call Access");
        break;
      case 404:
        toast.error("404 Not Found");
        break;
      case 500:
        toast.error("Server Is Down");
        break;
    }
    redirectToErrorPage()
    return Promise.reject(error);
  }
);
export default axiosInstance;
