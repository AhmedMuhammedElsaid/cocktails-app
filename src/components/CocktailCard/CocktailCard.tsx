import React, { FunctionComponent } from "react";
import {
  Card,
  CardActionArea,
  CardContent,
  CardMedia,
  Typography,
} from "@material-ui/core";
import styles from "./CocktailCard.module.scss";
import { cocktailSchema } from "utils/types";
import { Link } from "react-router-dom";
interface Props {
  cocktail: cocktailSchema;
  hideDesc?: boolean;
}

const CocktailCard: FunctionComponent<Props> = ({ cocktail, hideDesc }) => {
  const { strCategory, strDrink, strDrinkThumb, strInstructionsDE, idDrink } =cocktail;
  return (<>
          {/* 
        // @ts-ignore */}
          <Card component={Link} to={`/cocktail-details/${idDrink}`}
            className={styles.Card}
            data-testid="cocktail-card-testid"
          >
            <CardActionArea>
              <CardMedia
                className={styles.Card__Img}
                component="img"
                height="250"
                image={strDrinkThumb}
                alt={strDrink}
                data-testid="cocktail-img-testid"
              />
              <CardContent>
                <Typography
                  gutterBottom
                  variant="h6"
                  component="div"
                  data-testid="cocktail-name-testid"
                  className="Cocktail__Name"
                >
                  {strDrink}
                </Typography>
                {!hideDesc && (
                  <>
                    {" "}
                    <Typography
                      gutterBottom
                      variant="h6"
                      component="div"
                      className="Category"
                      data-testid="cocktail-category-testid"
                    >
                      {strCategory}
                    </Typography>
                    <Typography
                      variant="body2"
                      className="Description"
                      data-testid="cocktail-description-testid"
                    >
                      {strInstructionsDE?.slice(0, 120)}...
                    </Typography>
                  </>
                )}
              </CardContent>
            </CardActionArea>
          </Card>
          </>
  );
};
export default CocktailCard;
