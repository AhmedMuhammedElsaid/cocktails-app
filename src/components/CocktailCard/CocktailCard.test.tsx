import { cocktailCardMockedData } from "mocks/mockedData";
import { customRender } from "utils/testUtils";
import CocktailCard from "./CocktailCard";

const cocktailName = "cocktail-name-testid",
  cocktailImg = "cocktail-img-testid",
  cocktailCategory = "cocktail-category-testid",
  cocktailDesc = "cocktail-description-testid",
  noCocktailData = "no-cocktail-testid";
describe("Test Cocktail Card Item to render without Errors", () => {
  it("render card item details without crashing", () => {
    const { asFragment, getByTestId, queryByTestId } = customRender(
      <CocktailCard cocktail={cocktailCardMockedData} />
    );
    expect(queryByTestId(noCocktailData)).not.toBeInTheDocument();
    expect(getByTestId(cocktailName)).toHaveTextContent(
      cocktailCardMockedData.strDrink
    );
    expect(getByTestId(cocktailCategory)).toHaveTextContent(
      cocktailCardMockedData.strCategory
    );
    expect(getByTestId(cocktailDesc)).toHaveTextContent(
      cocktailCardMockedData.strInstructionsDE
    );
    expect(getByTestId(cocktailImg)).toHaveAttribute(
      "src",
      cocktailCardMockedData.strDrinkThumb
    );

    // take a snapshot to the component in order to have an image
    // to compare with future changes image
    expect(asFragment()).toMatchSnapshot();
  });
});
