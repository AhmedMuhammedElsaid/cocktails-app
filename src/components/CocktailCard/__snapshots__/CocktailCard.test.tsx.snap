// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cocktail Card Item to render without Errors render card item details without crashing 1`] = `
<DocumentFragment>
  <a
    class="MuiPaper-root MuiCard-root Card MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="cocktail-card-testid"
    href="/cocktail-details/undefined"
  >
    <button
      class="MuiButtonBase-root MuiCardActionArea-root"
      tabindex="0"
      type="button"
    >
      <img
        alt="Cream Soda"
        class="MuiCardMedia-root Card__Img MuiCardMedia-media MuiCardMedia-img"
        data-testid="cocktail-img-testid"
        height="250"
        src="https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg"
      />
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiTypography-root Cocktail__Name MuiTypography-h6 MuiTypography-gutterBottom"
          data-testid="cocktail-name-testid"
        >
          Cream Soda
        </div>
         
        <div
          class="MuiTypography-root Category MuiTypography-h6 MuiTypography-gutterBottom"
          data-testid="cocktail-category-testid"
        >
          Cocktail
        </div>
        <p
          class="MuiTypography-root Description MuiTypography-body2"
          data-testid="cocktail-description-testid"
        >
          Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale ....
        </p>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </a>
</DocumentFragment>
`;
